helios-system-tests/src/main/java/com/spotify/helios/system/HealthCheckTest.java

```diff
@@ -0,0 +1,220 @@
+import com.spotify.helios.serviceregistration.ServiceRegistrar;
+import com.spotify.helios.serviceregistration.ServiceRegistration;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import java.io.IOException;
+import java.net.Socket;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import static com.spotify.helios.common.descriptors.HostStatus.Status.UP;
+import static com.spotify.helios.common.descriptors.TaskStatus.State.RUNNING;
+import static com.spotify.helios.serviceregistration.ServiceRegistration.Endpoint;
+import static java.util.Arrays.asList;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.verify;
+
+@RunWith(MockitoJUnitRunner.class)
+public class HealthCheckTest extends ServiceRegistrationTestBase {
+
+  @Mock
+  public ServiceRegistrar registrar;
+
+  @Captor
+  public ArgumentCaptor<ServiceRegistration> registrationCaptor;
+
+  final String registryAddress = uniqueRegistryAddress();
+
+  @Before
+  public void setup() {
+    MockServiceRegistrarRegistry.set(registryAddress, registrar);
+  }
+
+  @After
+  public void teardown() {
+    MockServiceRegistrarRegistry.remove(registryAddress);
+  }
+
+  @Test
+  public void testTcp() throws Exception {
+    startDefaultMaster();
+
+    final HeliosClient client = defaultClient();
+
+    startDefaultAgent(testHost(), "--service-registry=" + registryAddress);
+    awaitHostStatus(client, testHost(), UP, LONG_WAIT_SECONDS, SECONDS);
+
+    final HealthCheck healthCheck = TcpHealthCheck.of("health");
+
+    // start a container that listens on a poke port, and once poked listens on the healthcheck port
+    final Job job = Job.newBuilder()
+        .setName(testJobName)
+        .setVersion(testJobVersion)
+        .setImage(BUSYBOX)
+        .setCommand(asList("sh", "-c", "nc -l -p 4711 && nc -l -p 4712"))
+        .addPort("poke", PortMapping.of(4711))
+        .addPort("health", PortMapping.of(4712))
+        .addRegistration(ServiceEndpoint.of("foo_service", "foo_proto"), ServicePorts.of("health"))
+        .setHealthCheck(healthCheck)
+        .build();
+
+    final JobId jobId = createJob(job);
+    deployJob(jobId, testHost());
+    awaitTaskState(jobId, testHost(), RUNNING);
+
+    // wait a few seconds to see if the service gets registered
+    Thread.sleep(3000);
+    // shouldn't be registered, since we haven't poked it yet
+    verify(registrar, never()).register(any(ServiceRegistration.class));
+
+    // poke container to get it to start listening on the healthcheck port
+    Polling.await(LONG_WAIT_SECONDS, SECONDS, new Callable<Object>() {
```

Comment:
I'm just trying to understand this test. So we poll the "poke" port until we can make a TCP connection to it. Then we check that `registrar.register()` has been called within `LONG_WAIT_SECONDS` because that means the health check was executed and succeeded?